flatbuffers

jmillan opened the pull request:
WIP to remove JSON in favour of [flatbuffers](https://github.com/google/flatbuffers) for as the worker communication format.

* JSON parsing is CPU intensive.
* JSON is not type safe. 
* flatbuffers does not parse the buffer. Reading the message takes 0ms.
* flatbuffers is type safe since code (C++, Typescript, Rust, etc) is autogenerated out of the schema definitions.

Extras that come along with this changes:

* There is now a single Channel for worker communication. 
  * Previously there were 2 (Channel and PayloadChannel)
* Each message sent Node->Worker requires a single write() call to Channel.
  * Previously 2 calls per message were needed for Channel and 4 for PayloadChannel.
* Each message sent Worker->Node previously executed two extra calls to `memcpy` that are not executed now.

### TODO: flatbuffers related
- [ ] Try to scope notifications per instance so we don't rely on a global flatbuffers notification event enum but an enum per class (producer, consumer, transport, etc.):  https://github.com/versatica/mediasoup/pull/1162 (non blocker)
- [ ] Do not use heap on flatbuffers generation (non blocker)
- [ ] Do not use head on flatbuffers processing (non blocker)

### TODO: flatbuffers unrealed. Changes made in flatbuffer branch, not present in v3
- [x] Document new `TransportListenInfo` in the website (PR https://github.com/versatica/mediasoup/pull/1084).
- [x] Document new `rtcpListenInfo` in `PlainTransportOptions` once PR https://github.com/versatica/mediasoup/pull/1099 is merged here.
- [x] Document the new pause/resume feature in `DataProducer` and `DataConsumer` (PR https://github.com/versatica/mediasoup/pull/1104):
  - `pause()`and `resume()` in `DataProducer` and `DataConsumer`.
  - `paused` getter in `DataProducer` and `DataConsumer`.
  - `dataproducerpause` and `dataproducerresume` events in `DataConsumer`.
  - `pause` and `resume` events in `DataProducer.observer` and `DataConsumer.observer`.
  - `paused?: boolean` in `transport.consumeData()` and `transport.produceData()`.
- [x] Document the new subchannels stuff (PR https://github.com/versatica/mediasoup/pull/1152).

jmillan committed:
flatbuffers

nazar-pc committed:
Rust: WIP

ibc commented:
@jmillan this branch uses NPM `flatbuffers` 23.3.3 and there is 23.5.9 already. CHANGELOG here: https://github.com/google/flatbuffers/blob/master/CHANGELOG.md

I'm merging `v3` into `flatbuffers` and will later upgrade this dep.

ibc committed:
Merge branch 'v3' into flatbuffers

# Conflicts:
#	worker/Makefile

ibc committed:
Update NPM flatbuffers dep

ibc commented:
> @jmillan this branch uses NPM `flatbuffers` 23.3.3 and there is 23.5.9 already. CHANGELOG here: https://github.com/google/flatbuffers/blob/master/CHANGELOG.md
> 
> I'm merging `v3` into `flatbuffers` and will later upgrade this dep.

Done and pushed.

ibc committed:
Merge branch 'v3' into flatbuffers

# Conflicts:
#	worker/src/RTC/RtpStream.cpp

ibc committed:
[flatbuffers] Add ListenInfo in all transports and send/recv buffer size options (#1084)

ibc committed:
Merge branch 'v3' into flatbuffers

ibc committed:
Merge branch 'v3' into flatbuffers

# Conflicts:
#	.github/workflows/mediasoup-node.yaml
#	.gitignore
#	CHANGELOG.md
#	npm-scripts.js
#	worker/src/lib.cpp

ibc committed:
Merge branch 'v3' into flatbuffers

# Conflicts:
#	CHANGELOG.md
#	npm-scripts.mjs
#	package.json

ibc committed:
Merge branch 'v3' into flatbuffers

# Conflicts:
#	npm-scripts.mjs

ibc commented:
Questions. Sometimes I see this:

```c++
if (flatbuffers::IsFieldPresent(options->listenInfo(), FBS::Transport::ListenInfo::VT_ANNOUNCEDIP))
```

and sometimes I see this (which is way nicer):

```c++
if (body->port().has_value())
{
    port = body->port().value();
}
```

I'd love if the latter could be used everywhere, no idea if possible.

ibc commented:
I may be crazy but `make lint` doesn't detect wrong changes in `PlainTransport.cpp` although it does in other `XxxxTransport.cpp` classes...

ibc commented:
@jmillan, in the TODO items above:

> flatbuffers (TS) should throw `TypeError` instead of `Error` when there is a missing required field. [PR done](https://github.com/google/flatbuffers/pull/7910).

It's been merged. Not sure if there is a new release of `flatbuffers` for Node, but if so we should change some code, right?

ibc committed:
PlainTransportOptions: add optional rtcpListenInfo (#1099)

jmillan committed:
Remove explicit TypeError exceptions

Already thown by flatbuffers.

jmillan commented:
>> flatbuffers (TS) should throw TypeError instead of Error when there is a missing required field. https://github.com/google/flatbuffers/pull/7910.

> It's been merged. Not sure if there is a new release of flatbuffers for Node, but if so we should change some code, right?

Removed explicit `TypeError` exceptions.

ibc committed:
Merge branch 'v3' into flatbuffers

ibc committed:
[flatbuffers] Add pause() and resume() in DataProducer and DataConsumer (#1104)

ibc committed:
Merge branch 'v3' into flatbuffers

# Conflicts:
#	CHANGELOG.md

ibc committed:
Merge branch 'v3' into flatbuffers

ibc committed:
Merge branch 'v3' into flatbuffers

ibc committed:
Merge branch 'v3' into flatbuffers

# Conflicts:
#	CHANGELOG.md

ibc committed:
Merge branch 'v3' into flatbuffers

# Conflicts:
#	CHANGELOG.md

ibc committed:
Merge branch 'v3' into flatbuffers

ibc committed:
Merge branch 'v3' into flatbuffers

ibc committed:
Merge branch 'v3' into flatbuffers

ibc committed:
Merge branch 'v3' into flatbuffers

# Conflicts:
#	worker/src/RTC/RtpStreamRecv.cpp

jmillan committed:
Merge remote-tracking branch 'origin/flatbuffers' into flatbuffers-rust-wip

jmillan committed:
rust: remove unneeded file

jmillan committed:
rust: disable tests for unimplemented code

And enable 'cargo test' workflow in GH actions.

ibc committed:
Merge branch 'v3' into flatbuffers

ibc committed:
cosmetic

jmillan commented:
@ibc, the `DataConsumer.[pause|consumer]()` implementation is making the test fail in GH actions. 

Calling `pause()` or `resume()` is sending the corresponding method to the worker and only after receiving the response it's setting the local `paused` member.

The test just checks if `dataConsumer.paused` in order to increase a variable in the test `effectivelySentMessageBytes`, but it can happen that the `pause()` method has been sent but not terminated, so the Node instance `paused` member is false.

I believe that can be the problem, and probably it's not an error per-se but something to be fixed in the test. I was able to repro the error just once, locally:

`TEST_FILE=node/src/tests/test-DirectTransport.ts npm run test:node`

<img width="976" alt="Screenshot 2023-07-17 at 13 03 55" src="https://github.com/versatica/mediasoup/assets/732340/7731e331-0014-41c2-90e0-65a6b33e00cb">

ibc committed:
Merge branch 'v3' into flatbuffers

ibc commented:
Specially in flatbuffers branch, requests and data messages are sent over the same channel so it's guaranteed that, if I pause the DataProducer and then send a message, such a message won't reach the DataConsumers. I'm 95% sure I took this into account when I wrote the corresponding tests. So can we agree on these things?

1. Code is good and behaves as expected.
2. So maybe tests are wrongly written.

ibc commented:
I think the problem could he here:

![CleanShot-2023-07-17-at-13 41 50@2x](https://github.com/versatica/mediasoup/assets/16191/7c780bcc-4269-4d4d-a504-f8a4f07c57db)

We are setting the `paused = true` flag in `pause()` method ONCE we get ok from the worker. However in the test we do NOT await for `dataProducer.pause()` to resolve, and in lines below the test checks `dataProducer.paused` to decide things.

ibc commented:
It happened to me too. I'm working on it.
<img width="576" alt="CleanShot-2023-07-17-at-13 45 15@2x" src="https://github.com/versatica/mediasoup/assets/16191/aba41813-3307-4d8f-bd27-fa32ff800080">

ibc commented:
@jmillan does it make sense that `dataProducer.pause()` sends a request to the worker (same for `Dataconsumer` and same for `resume()` method in both? For instance, `dataProducer.send()` sends a **notification** to the worker. We have experience on this. Sending a request means that `dataConsumerpause()` will throw if the `DataConsumer` is closed and forces the user to always add `catch()` etc.

I think it's a good timing to refactor those requests and make them notifications. On it.

ibc commented:
> @jmillan does it make sense that `dataProducer.pause()` sends a request to the worker (same for `Dataconsumer` and same for `resume()` method in both? For instance, `dataProducer.send()` sends a **notification** to the worker. We have experience on this. Sending a request means that `dataConsumerpause()` will throw if the `DataConsumer` is closed and forces the user to always add `catch()` etc.
> 
> I think it's a good timing to refactor those requests and make them notifications. On it.

It's not a good timing for doing this. Let's do this in v4. Milestone created in GH and first item:

https://github.com/versatica/mediasoup/issues/1121

ibc committed:
Fix DirectTransport test when using pause/resume()

As explained in https://github.com/versatica/mediasoup/pull/1064#issuecomment-1637902087 and messages below it

ibc commented:
@jmillan I've pushed the fix: https://github.com/versatica/mediasoup/pull/1064/commits/40379354cb60634c4051f5c2a9657b5964211aa5
Interesting thing is that the corresponding Rust test was properly done from the beginning (it awaits for puase() and resume() to resolve).

ibc committed:
Merge branch 'v3' into flatbuffers

# Conflicts:
#	worker/include/RTC/PipeConsumer.hpp

ibc committed:
Merge branch 'v3' into flatbuffers

# Conflicts:
#	CHANGELOG.md

piranna commented:
How does `flatbuffers` compare in memory and performance to other binary alternatives like [CBOR-X](https://github.com/kriszyp/cbor-x)? I though it was the faster and most memory efficient...

ibc commented:
There is no reason to compare flatbuffers (available in JS, C++, Rust) with something that is just for JS (so not useful for us).

piranna commented:
I'm reading about Flatbuffers and looks really interesting, and seems to be on pair in both memory and performance. I've open an issue at https://github.com/google/flatbuffers/issues/8046 asking for some benchmark metrics comparing both of them.

threema-lenny commented:
> I'm reading about Flatbuffers and looks really interesting, and seems to be on pair in both memory and performance. I've open an issue at [google/flatbuffers#8046](https://github.com/google/flatbuffers/issues/8046) asking for some benchmark metrics comparing both of them.

It makes no sense to compare the two in this particular use case. CBOR needs serialisation and deserialisation whereas FlatBuffers does not. The winner is therefore already known. Everything else is an implementation detail. They target totally different use cases. Using FlatBuffers for IPC is a very sensible approach. CBOR would introduce similar overhead to JSON.

piranna commented:
So, maybe CBOR can be a advantage on the bandwidth usage, for example by sending small numbers as varlen instead of full ints, isn't It?

ibc commented:
Can we please avoid discussing in this PR about different solutions to the same problem? Flatbuffers was chosen after long research and we are not gonna change it. Also, I already said that we need something that works in Node, Rust and C++. Flatbuffers does. CBOR is only for JS.

nazar-pc commented:
> So, maybe CBOR can be a advantage on the bandwidth usage, for example by sending small numbers as varlen instead of full ints, isn't It?

We don't care about bandwidth usage at all, please don't distract us with irrelevant discussions here. Feel free to start a topic on the forum though.

piranna commented:
I'm not discussing about a different solution, just wanted to know about why was this one choosen, definitely it seems to be one of the best options, just I didn't know about it. Sorry for the inconvenience.

ibc commented:
It was chosen because it's available in all languages we need, it has very good performance (specially compared to parsing and serializing JSON in C++ using JsonCpp lib, Jose did benchmarks), and doesn't require any memory allocation (well, just a global buffer) since it writes and reads in place.

nazar-pc committed:
Update planus and re-generate types

nazar-pc committed:
Implement `RtpParameters::from_fbs()` and `RtpParameters::into_fbs()`

nazar-pc committed:
Commit `Cargo.lock` file like it is done for Node.js

nazar-pc committed:
Merge pull request #1131 from versatica/flatbuffers-rtp-parameters

Implement RtpParameters::from_fbs() and RtpParameters::into_fbs()

ibc committed:
Merge branch 'v3' into flatbuffers

jmillan committed:
FBS enhancements (#1134)

* RtpParameters: make Parameter.value required

* RtpParameters: make String Parameter required

* RtpParameters: make RtcpFeedback parameter required

* RtpParameters: make encodings required

* RtpParameters: make RtcpParameters required

* Rust: RtpEncodingParameters, remove scale_resolution_by

* fix tests

* Rust: RtpParameters, make headerExtensions required

* Worker: consider new required RtpParameters

ibc committed:
Merge branch 'v3' into flatbuffers

ibc committed:
Merge branch 'v3' into flatbuffers

# Conflicts:
#	CHANGELOG.md
#	node/src/Router.ts

ibc committed:
Merge branch 'v3' into flatbuffers

# Conflicts:
#	CHANGELOG.md

ibc committed:
Merge branch 'v3' into flatbuffers

ibc committed:
Merge branch 'v3' into flatbuffers

# Conflicts:
#	CHANGELOG.md

jmillan committed:
rust: move PlainTransport to flatbuffers (#1135)

jmillan committed:
Use FBS Transport::Tuple everywhere (#1148)

jmillan committed:
rust: move PipeTransport to flatbuffers (#1149)

jmillan committed:
Merge remote-tracking branch 'origin/v3' into flatbuffers

ibc committed:
Merge branch 'v3' into flatbuffers

# Conflicts:
#	CHANGELOG.md

Commit:
merge

ibc committed:
DataChannel subchannels (#1152)

ibc commented:
So regarding the Rust `ntapi` issue:

> > https://github.com/versatica/mediasoup/actions/runs/6121294807/job/16619834171?pr=1152
> > ```
> > error[E0793]: reference to packed field is unaligned
> >     --> C:\Users\runneradmin\.cargo\registry\src\index.crates.io-6f17d22bba15001f\ntapi-0.3.7\src\ntexapi.rs:2783:52
> >      |
> > 2783 |         *tick_count.QuadPart_mut() = read_volatile(&(*USER_SHARED_DATA).u.TickCountQuad);
> >      |                                                    ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
> >      |
> >      = note: packed structs are only aligned by one byte, and many modern architectures penalize unaligned field accesses
> >      = note: creating a misaligned reference is undefined behavior (even if that reference is never dereferenced)
> >      = help: copy the field contents to a local variable, or replace the reference with a raw pointer and use 
> > ```
> 
> So there is a related bug: https://bugzilla.mozilla.org/show_bug.cgi?id=1826485
> 
> It seems to be fixed in `ntapi` 0.4.0: https://bugzilla.mozilla.org/show_bug.cgi?id=1826485#c2
> 
> However I have no idea which `ntapi` version we are using (whatever `ntapi` is). Output of `cargo tree` doesn't show `ntapi` at all.
> 
> Then people say here that _"For some reason (I don't know why) Cargo picked an older version of tokio that used (indirectly) an older version of ntapi. Force Cargo to use new tokio by setting its version to "1.28.0" (the newest version), and it should work."_. We have `tokio` 1.17.0 (whatever `tokio` is) according to `Cargo tree`. But of course we don't specify `tokio` in any `Cargo.toml` so...

@nazar-pc can we do this in `mediasoup/rust/Cargo.toml`?:
<img width="759" alt="Screenshot 2023-09-11 at 12 27 04" src="https://github.com/versatica/mediasoup/assets/16191/a166cb46-4720-4f29-9f5e-06c59b07f620">

ibc committed:
Update CHANGELOG

nazar-pc commented:
> @nazar-pc can we do this in mediasoup/rust/Cargo.toml?:

Just do `cargo update -p tokio` and it'll be updated in lock file. Or `cargo update` to update all dependencies at once.

ibc committed:
[flatbuffers] Update rust deps (#1155)

ibc committed:
Update FBS for Rust

As Nazar said, for now the process is manual:

1. In `rust` directory run this: `planus rust -o src/fbs.rs ../worker/fbs/*`
2. Then revert some of the changes at the top of the file to make sure it re-exports things the same way and suppresses clippy warnings in there.
3. For ^ to work you might need to install it first with `cargo install planus-cli` (this will download, compile and install binary in Cargo's directory).

ibc committed:
cosmetic: add blank line

jmillan committed:
fbs: FBS WebRtcTransport (#1153)

fbs: FBS WebRtcTransport

---------

Co-authored-by: Iñaki Baz Castillo <ibc@aliax.net>

ibc committed:
Make DTLS fragment stay within MTU size range (#1156)

jmillan committed:
Rust: port transport methods to FBS (#1157)

Rust: port transport methods to FBS

ibc committed:
Jest coverage: ignore tests/ and fbs/ folders

jmillan committed:
Rust: port transport notifications to FBS (#1158)

Rust: port transport notifications to FBS

jmillan committed:
FBS: normalize FBS types (#1159)

* FBS: normalize FBS types

* FBS: format schemas

* FBS: type for RtpHeaderExtensionUri (#1161)

ibc committed:
Merge branch 'v3' into flatbuffers

# Conflicts:
#	CHANGELOG.md
#	worker/Makefile

jmillan committed:
FBS: Producer (#1164)

FBS: Producer

jmillan committed:
FBS: Consumer (#1165)

FBS: Consumer

jmillan committed:
FBS: Producer notifications (#1166)

jmillan committed:
FBS: RTP observers

jmillan committed:
FBS: DataProducer, DataConsumer (#1169)

jmillan committed:
Rust: uncomment remanning ignored test (#1170)

Few tests already commented since they are being worked out.

jmillan committed:
Rust: remove temporal FBS suffixes (#1171)

They were temporal until every message was ported to flatbuffers.

ibc committed:
Merge branch 'v3' into flatbuffers

jmillan committed:
Rust: properly store incoming notifications (#1172)

Store the whole binary message and retrieve the notification reference before
calling the handler.

jmillan committed:
Rust: add missing request (#1173)

Bonus: fix test.

jmillan commented:
Update: after running the current benchmarks in Rust (only available for dataproducer.send()), I see an overall performance increase of 15% comparing to v3.

I've not done any other performance test. It would be interesting to write some test implying more transformations to from FBS.

jmillan committed:
Rust: messages, fix DataProducerSendNotification

nazar-pc commented:
Will probably get way better once we stop allocating right and left in Rust code (pretty much everything is boxed right now, sometimes a few times).

ibc committed:
Merge branch 'v3' into flatbuffers

# Conflicts:
#	CHANGELOG.md

ibc commented:
There is a bug. I run mediasoup `flatbuffers` in the demo with `DEBUG="*"` and I see this wrong "unexpected" output:

```
  mediasoup:Worker constructor() +0ms
  mediasoup:Worker spawning worker process: /Users/ibc/src/v3-mediasoup/worker/out/Release/mediasoup-worker --logLevel=debug --logTag=info --logTag=ice --logTag=dtls --logTag=rtp --logTag=srtp --logTag=rtcp --logTag=rtx --logTag=bwe --logTag=score --logTag=simulcast --logTag=svc --logTag=sctp --rtcMinPort=40000 --rtcMaxPort=49999 +0ms
  mediasoup:Channel constructor() +0ms
  mediasoup:PayloadChannel constructor() +0ms
worker[pid:77521] unexpected data:


_Dmediasoup-worker::mediasoup_worker_run() | starting mediasoup-worker process [version:3.12.13]
worker[pid:77521] unexpected data:


FDmediasoup-worker::mediasoup_worker_run() | little-endian CPU detected
worker[pid:77521] unexpected data:


IDmediasoup-worker::mediasoup_worker_run() | 64 bits architecture detected
worker[pid:77521] unexpected data:


1DSettings::PrintConfiguration() | <configuration>
worker[pid:77521] unexpected data:


@DSettings::PrintConfiguration() |   logLevel             : debug
worker[pid:77521] unexpected data:


wDSettings::PrintConfiguration() |   logTags              : info,ice,dtls,rtp,srtp,rtcp,rtx,bwe,score,simulcast,svc,sctp
worker[pid:77521] unexpected data:


@DSettings::PrintConfiguration() |   rtcMinPort           : 40000
worker[pid:77521] unexpected data:


@DSettings::PrintConfiguration() |   rtcMaxPort           : 49999
worker[pid:77521] unexpected data:


`DSettings::PrintConfiguration() |   libwebrtcFieldTrials : WebRTC-Bwe-AlrLimitedBackoff/Enabled/
worker[pid:77521] unexpected data:


2DSettings::PrintConfiguration() | </configuration>
worker[pid:77521] unexpected data:


3DDepLibUV::PrintVersion() | libuv version: "1.44.2"
worker[pid:77521] unexpected data:


GDDepOpenSSL::operator()() | openssl version: "OpenSSL 3.0.8 7 Feb 2023"
worker[pid:77521] unexpected data:


<DDepLibSRTP::ClassInit() | libsrtp version: "libsrtp2 2.5.0"

[opening Readline Command Console...]
type help to print available commands
cmd> worker[pid:77521] unexpected data:


"DDepUsrSCTP::ClassInit() | usrsctp
worker[pid:77521] unexpected data:


\DDepLibWebRTC::ClassInit() | libwebrtc field trials: "WebRTC-Bwe-AlrLimitedBackoff/Enabled/"
worker[pid:77521] unexpected data:


tDwebrtc::FieldTrial::InitFieldTrialsFromString() | Setting field trial string: WebRTC-Bwe-AlrLimitedBackoff/Enabled/
worker[pid:77521] unexpected data:


�DRTC::DtlsTransport::CreateSslCtx() | setting SRTP cryptoSuites for DTLS: SRTP_AEAD_AES_256_GCM:SRTP_AEAD_AES_128_GCM:SRTP_AES128_CM_SHA1_80:SRTP_AES128_CM_SHA1_32
worker[pid:77521] unexpected data:


�DRTC::DtlsTransport::GenerateFingerprints() | sha-224 fingerprint: DF:1B:4A:DD:8B:B7:F9:2D:4D:EE:25:06:F0:B5:A1:27:C3:A4:C0:E5:65:17:2B:EB:46:D0:C9:AD
worker[pid:77521] unexpected data:


�DRTC::DtlsTransport::GenerateFingerprints() | sha-256 fingerprint: BA:2D:F8:86:1F:77:28:BD:10:56:93:F8:A0:01:04:05:3A:70:A8:D2:07:BD:75:36:9C:F5:B2:2B:1C:F1:7D:2D
worker[pid:77521] unexpected data:


~DRTC::DtlsTransport::GenerateFingerprints() | sha-1   fingerprint: 7F:48:8A:6F:08:78:CB:19:15:28:0A:79:7C:FE:7C:FC:59:28:D9:3D
worker[pid:77521] unexpected data:


�DRTC::DtlsTransport::GenerateFingerprints() | sha-384 fingerprint: 9F:79:AA:43:DE:42:53:50:3A:A9:EA:E5:02:17:16:87:27:F4:AE:29:58:EE:72:6C:74:15:84:F2:4A:B5:D4:9E:19:37:4B:5B:A0:C2:29:9C:8E:4F:D1:3D:70:9C:23:B3
worker[pid:77521] unexpected data:


DRTC::DtlsTransport::GenerateFingerprints() | sha-512 fingerprint: 1E:09:89:46:B0:36:D9:B9:0A:B5:66:FF:1B:63:C2:DC:92:95:58:62:FB:61:38:A9:26:52:F4:23:48:32:6B:D5:14:55:6B:25:38:2D:45:46:3D:C9:37:31:CE:C9:5D:A7:11:DB:03:49:8D:B6:8D:B4:3A:DF:5D:AD:2A:EC:1C:44
worker[pid:77521] unexpected data:




77521
```

Note that it's not "unexpcted". It's data that the worker must send to Node. So it looks like if it's being properly received by Node but then Node is also printing it to the console (no idea, it could be another thing).

ibc commented:
**BUG:**

```
webRtcServer.dump()
ERROR: webRtcServer.dump() failed: TypeError: Do not know how to serialize a BigInt
```

### STR:

- Run the demo and start a meeting.
- In the interactive console in the server, run `dwrs`.

I've added this item to the TODO list in this PR description.

ibc commented:
Question: is there any reason for all `dump()` methods to still return `Promise<any>`? We do have those returned TS types defined since they are FBS based, right?

ibc commented:
> **BUG:**
> 
> ```
> webRtcServer.dump()
> ERROR: webRtcServer.dump() failed: TypeError: Do not know how to serialize a BigInt
> ```

I've created a separate ticket for this: https://github.com/versatica/mediasoup/issues/1180

ibc committed:
Fix webRtcServer.dump()

Fixes #1180

jmillan commented:
> Question: is there any reason for all dump() methods to still return Promise<any>? We do have those returned TS types defined since they are FBS based, right?

They are supposed to be typed already https://github.com/versatica/mediasoup/blob/flatbuffers/node/src/WebRtcTransport.ts#L438

If you find any untyped please type it or add a TODO item on this PR.

ibc commented:
> > Question: is there any reason for all dump() methods to still return Promise<any>? We do have those returned TS types defined since they are FBS based, right?
> 
> 
> 
> They are supposed to be typed already https://github.com/versatica/mediasoup/blob/flatbuffers/node/src/WebRtcTransport.ts#L438
> 
> 
> 
> If you find any untyped please type it or add a TODO item on this PR.

I meant WebRtcServer.dump() and I've seen others. I'll create a separate issue.

ibc commented:
I've created a separate issue: 
Must add TS return type to all dump() methods: https://github.com/versatica/mediasoup/issues/1181

jmillan committed:
Enhance flatc build system (#1174)

Meson: properly generate C++ flatbuffers header files

- Do only generate them when the flatbuffer schemas change
- Make generated FBS files a worker dependency
- Create a meson.build for 'fbs' directory
- Remove _generated suffix to generated fbs files

jmillan commented:
> Must add TS return type to all dump() methods: https://github.com/versatica/mediasoup/issues/1181

@ibc, I don't agree with this TODO list here. It could have been perfectly typed before, as most of responses are, generally `dump()` responses have not been typed nor documented on purpose. This is not a blocking task for this branch, but can be something to do anytime. This branch already has too many working hours and I don't won't to add extra weight on it.

jmillan commented:
> Be ready to serialize BigInt in Node

@ibc, what are we supposed to do here? We are casting every BigInt to Number already. I don't think this is a blocker for this branch.

ibc commented:
> generally dump() responses have not been typed nor documented on purpose. This is not a blocking task for this branch

Agreed, removed from TODO. It's just a separate issue now.

ibc commented:
> We are casting every BigInt to Number already. I don't think this is a blocker for this branch.

Are we? I just found a missing case yesterday. I'm not saying that there are more, but are we sure? The error I fixed yesterday just happens when a dump (or whatever response or notification) contains a super big number. It doesn't fail in build time but just in runtime in certain cases. Are we sure we are casting to Number absolutely all fields defined as uint64 or int65 in FBS? If so we are good.

ibc commented:
I've added a blocker: https://github.com/versatica/mediasoup/issues/1184

ibc committed:
Merge branch 'v3' into flatbuffers

# Conflicts:
#	CHANGELOG.md

jmillan committed:
Fbs RTP trace info (#1183)

FBS: RTP trace info

Also fix flatbuffers include statement order

ibc committed:
[flatbuffers] Fix missing options in router.createWebRtcTransport() (#1185)

jmillan committed:
C++: Fix DataProducer::Send invalid memory access (#1190)

'subchannels' are optional and hence they need to be checked before
accessing them.

Remove test expectation that assumed we were receiving all messages in
incremental mode. TS tests failure was being masked by [1], I don't know
why Rust wasn't failing but definitely consumers are not receiving
everything sent by the producer since in the tests both are paused at
some time.

[1]: https://github.com/versatica/mediasoup/issues/1188

ibc committed:
Merge branch 'v3' into flatbuffers

# Conflicts:
#	CHANGELOG.md

ibc committed:
Merge branch 'v3' into flatbuffers

# Conflicts:
#	CHANGELOG.md

jmillan committed:
Rust: Fix worker error handling

Properly retrieve the error reason.

jmillan committed:
Rust: Data[Producer|Consumer] subchannels (#1191)

Rust: Data[Producer|Consumer] subchannels

jmillan commented:
@nazar-pc, time has come for the first two TODO items:

 - [ ] Build fbs.rs at build time (planus).
 - [ ] Format fbs schemas at build time (planus).

nazar-pc commented:
What do you mean by formatting?

jmillan commented:
> What do you mean by formatting?

Planus CLI has a ' format' option that formats the flatbuffers schemas files. We should have this done the same way we run `cargo fmt` on the rest of the code.

nazar-pc commented:
Can you run it manually or you want to check in CI, what is the goal?

jmillan commented:
> Can you run it manually or you want to check in CI, what is the goal?

We can run it manually. No need to check in CI since it just overrides current files properly formatted. I'm removing from the TODO list.

jmillan commented:
About `Build fbs.rs at build time (planus).` task. It's not a blocker for me. It is a single autogenerated file what Planus outputs from all schemas so I'm happy generating it with `planus-cli` and keeping in the repo as we do now.

nazar-pc commented:
I will work on that, but probably closer to the week end, very busy

ibc commented:
> About `Build fbs.rs at build time (planus).` task. It's not a blocker for me. It is a single autogenerated file what Planus outputs from all schemas so I'm happy generating it with `planus-cli` and keeping in the repo as we do now.

Please let's not release without this. Having to rewrite those things in `fbs.rs` after every change in any FBS file is tedious and it would just be impossible for any other current or future contributor.

We need a proper command to generate `fbs.rs` without having to manually edit it after it. BTW should this be a new task in `worker/Makefile` or somewhere else? Can we define a custom `cargo run planus` command or similar? Does such a thing exist in Cargo land?


> "Format fbs schemas at build time (planus)"

What does this mean? Format which fbs schemas? You mean the source FBS files in `worker/fbs/` folder? Why do we want that those files are autoformatted by planus?

jmillan commented:
> What does this mean? Format which fbs schemas? You mean the source FBS files in worker/fbs/ folder? Why do we want that those files are autoformatted by planus?

Planus is the only guy in town that has a formatter for .fbs files.

jmillan committed:
format fbs files

jmillan commented:
Some performance notes:

|| V3 | flatbuffers |
|-|----|-----------|
|Producer creation| 42us | 24us|
|Producer dump  | 30us | 14us|

Basically we can say we can now do twice as much operations as in v3 in the same time.

ibc committed:
Update NPM dev deps

ibc committed:
[flatbuffers] Improve Node imports (#1193)

- Add `node:` prefix to all built-in Node imports (replaces PR #1142).
- Replace `uuid` package with built-in `crypto.randomUUID()`(replaces PR #1140).
- Cosmetic changes in TS files:
  - Avoid unnecessary `const { createWorker } = mediasoup` in almost all tests.
  - Move all FBS related imports to the bottom.
  - Some formating fixes that bypassed our ESLint rules.

ibc committed:
Add Node 16 in CI workflows/mediasoup-node.yaml (since we do support Node 16)

jmillan committed:
Rust: add bench for producer creation/dump (#1192)

Rust: add bench for producer creation/dump

ibc committed:
Use `npm ci --force` to make CI work in Node 16

ibc commented:
@jmillan this TODO item in the PR description references a merged PR:

> - [ ] Try to scope notifications per instance so we don't rely on a global flatbuffers notification event enum but an enum per class (producer, consumer, transport, etc.): https://github.com/versatica/mediasoup/pull/1158/files#r1328919685 (non blocker)

Should be checked?

jmillan commented:
> @jmillan this TODO item in the PR description references a merged PR:
> 
> > * [ ]  Try to scope notifications per instance so we don't rely on a global flatbuffers notification event enum but an enum per class (producer, consumer, transport, etc.): https://github.com/versatica/mediasoup/pull/1158/files#r1328919685 (non blocker)
> 
> Should be checked?

The link was wrong, fixed.

jmillan committed:
Rust: FBS do not use heap to process responses (#1194)

* Rust: FBS do not use heap to process responses

Avoid heap allocations. Provides 10% of performance improvement[*].

This is a test bed for rewritting the rest of `from_fbs` to
`from_fbs_ref`.

NOTE: Once all `from_fbs` are moved to `from_fbs_ref`, rename the
methods back to `from_fbs`

[*]: benched producer::dump()

jmillan committed:
Node: Add RouterDump and DirectTransportDump types (#1195)

Node: Add missing method return types

ibc committed:
[flatbuffers] Fix wrong usage of fs.readdirSync() in Node 20 (#1197)

Same as https://github.com/versatica/mediasoup-client/pull/283

ibc committed:
[flatbuffers] Remove '__MEDIASOUP_VERSION__' replacement in JS transpiled files (#1198)

- Instead of writing into all transpiled JS files in `node/lib/`, read `package.json` just once in launch time and read "version" from it.
- I'm not happy with this, not sure if better than before. But clearly we are gonna remove this useless `mediasoup.version` public getter in v4.

ibc commented:
@jmillan @nazar-pc, the most important missing piece here is:

> - [ ] Build `rust/src/fbs.rs` at build time (planus) (blocker)

So I'd like to discuss about this.

### Problem

Currently, as soon as we change/add/delete any file in `worker/fbs` folder, we must perform these steps manually in `rust` directory:

1. We need to have `planus-cli` installed (`cargo install planus-cli`).
2. `planus rust -o src/fbs.rs ../worker/fbs/*.fbs`
3. Then revert some of the changes at the top of `rust/src/fbs.rs` file to make sure it re-exports things the same way and suppresses clippy warnings.

### Questions

1. Does an external contributor have to install `planus-cli` manually? Or are we gonna add `planus-cli` in `[dev-dependencies]` in `rust/Cargo.toml`?
  - Off-topic: How is the `cargo` command to install dev deps?

2. So assuming we have `planus-cli`, how are we gonna automatize steps 2 and 3 above? Is this gonna be a new task in main `Makefile`? Or can we have a custom `cargo run planus` command (does such a thing exist?) and use `cargo-cli` as a lib?

nazar-pc commented:
1. Neither, we'll use Planus as a library and compile files programmatically from `build.rs` instead
2. They are installed automatically when you run tests, etc. It is not the same notion as Node.js. For equivalent stuff to Node.js there are `[build-dependencies]`, those are dependencies needed for `build.rs`
3. We will programmatically compile files as part of `build.rs` and emit [`cargo:rerun-if-changed`](https://doc.rust-lang.org/cargo/reference/build-scripts.html#rerun-if-changed) to instruct Cargo to re-run `build.rs` only if any of the files change (should also include C++ files)

ibc committed:
art/mediasoup-v3-architecture-01.svg: rename PlainRtpTransport to PlainTransport

ibc committed:
Merge branch 'v3' into flatbuffers

ibc commented:
> we'll use Planus as a library and compile files programmatically from `build.rs` instead

- So we have to run `cargo build` or `cargo test` to update fbs.rs, right?
- So we have to add ugly code in `build.rs` to "revert some of the changes at the top of `rust/src/fbs.rs` file to make sure it re-exports things the same way and suppresses clippy warnings", right?

nazar-pc commented:
> So we have to run cargo build or cargo test to update fbs.rs, right?

Neither and both. We will no longer ship `fbs.rs`, it will be compiled automatically when you run `cargo test` or use `mediasoup` as a dependency, but as mentioned before, only if it wasn't compiled already or corresponding files have changed.

> So we have to add ugly code in build.rs to "revert some of the changes at the top of rust/src/fbs.rs file to make sure it re-exports things the same way and suppresses clippy warnings", right?

I believe we do not, we can just suppress warnings on the whole module. Basically we will have `#[allow(clippy::whatever)] mod fbs {included_code_here}` or something like that in `lib.rs`, no manual search/replace will be necessary in `build.rs`.

ibc committed:
Fix unfortunate name in worker/scripts/package.json

jmillan committed:
FBS: Data[Consumer|Producer]  Send() simplifyfication (#1199)

FBS: Simplify Data[Consumer|Producer] send messages

For some reason we were not sending everything as binary in this branch.

jmillan committed:
Rust: fix bench

Update Cargo.lock which was missing.

nazar-pc committed:
Compile flatbuffers for Rust in build.rs

nazar-pc committed:
Merge pull request #1202 from versatica/compile-flatbuffers-build.rs

Compile flatbuffers for Rust in build.rs

jmillan committed:
FBS: Remove unneeded message type member (#1203)

FBS: Remove unneeded message type member

It's not needed since the body contained in the message is aleady typed.

ibc committed:
gitignore .nvmrc file

ibc committed:
Update Node deps

ibc commented:
I'm writing all missing docs in https://github.com/versatica/mediasoup-website/pull/50

nazar-pc commented on .gitignore:
```diff
@@ -7,8 +7,10 @@
+# flatc generated files.
 /node/src/fbs
+# nvm version file.
+/.nvmrc
```
Why was this necessary? You can ignore it locally, I don't see why it is necessary in the repo.

ibc replied:
I cannot ignore locally (meaning in my global .gitignore) since there are projects that **indeed** want this file into the repo. Just wondering what is so wrong with this addition in .gitignore taking into account that there are many other similar files in there such as ".idea" and so on.

nazar-pc replied:
You can put whatever you want into `.git/info/exclude` of each repo, no need to put stuff not related to the project in `.gitignore`.

ibc replied:
Do you mean that such a "exclude" folder is per local repo (so you and me can have different things into it)?

BTW why did you never say anything about this? 😅

![image](https://github.com/versatica/mediasoup/assets/16191/f5af5803-9a60-4ebb-b15c-20d55e9aff1b)

nazar-pc replied:
> Do you mean that such a "exclude" folder is per local repo (so you and me can have different things into it)?

Exactly, I usually put `.idea` in there when contributing to various projects, I don't expect all of them to use or care about JetBrains's IDEs.

> BTW why did you never say anything about this? 😅

Maybe I should have :man_shrugging:

ibc replied:
Ok I'll write PR later. I'll remove all those items but /.cache and /coverage which are autogenerated when running certain tasks. What about .tgz? Are we generating something.tgz and leaving it in root folder?

nazar-pc replied:
Things that are created by certain tasks actually should be in there, stuff that have nothing to do with the repo itself should not.

piranna replied:
I also add `/*.tgz` to `.gitignore` file, and also on `.npmignore`.

ibc replied:
We don't need .npmignore because we are explicit about npm package files in `package.json` (we fill the "files" field). Also, if `.npmignore` is not preset it reads `.gitignore`.

ibc replied:
No need for a PR. I've done basically what is told above: https://github.com/versatica/mediasoup/pull/1064/commits/96f0f45a3d3478e4f1d815f84b06da9b8b0cff75

ibc committed:
Clean .gitignore

Do not put there anything that it's not related to the project (such as files autogenerated by project tasks/scripts).
Everything else (`.vscode`, `.idea`, `.DS_Store`, etc) can be individually added to our local `.git/info/exclude` file).

ibc commented:
@jmillan I'm marking this PR as ready.

ibc committed:
Transport.ts: ConsumerAppData => DataConsumerAppData

jmillan committed:
FBS: Node, Fix RtpParameters (#1205)

* FBS: Node, Fix RtpParameters

Always pass encodings, header extensions and rtcp.

jmillan committed:
clang-tidy: exclude cppcoreguidelines-avoid-do-while

It makes it complain for every log or assert macro.

ibc committed:
Improve npm-scripts (#1207)

- Do not call `cleanWorkerArtifacts()` in `flatcNode()`. No reason for that plus it is bypassing the `MEDIASOUP_LOCAL_DEV` env purpose.
- Use more readable object arguments.

ibc committed:
Remove GCC warnings (#1208)

Fixes #1206

jmillan committed:
FBS: Fix DtlsTransport regression (#1210)

Do not process handshake if there is no remote fingerprint.

ibc committed:
Update Node deps

ibc committed:
DtlsTransport.cpp: cosmetic, add braces in if/else and case statements
